library/core/src/array/mod.rs

```diff
@@ -463,6 +463,34 @@ impl<T, const N: usize> [T; N] {
         unsafe { crate::mem::transmute_copy::<_, [U; N]>(&dst) }
     }
 
+    /// 'Zips up' two arrays into a single array of pairs.
+    /// `zip()` returns a new array where every element is a tuple where the first element comes from the first array, and the second element comes from the second array.
+    /// In other words, it zips two arrays together, into a single one.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(array_zip)]
+    /// let x = [1, 2, 3];
+    /// let y = [4, 5, 6];
+    /// let z = x.zip(y);
+    /// assert_eq!(z, [(1, 4), (2, 5), (3, 6)]);
+    /// ```
+    #[unstable(feature = "array_zip", issue = "none")]
+    pub fn zip<U>(self, rhs: [U; N]) -> [(T, U); N] {
+        use crate::mem::MaybeUninit;
+
+        let mut dst = MaybeUninit::uninit_array::<N>();
+        for ((lhs, rhs), dst) in IntoIter::new(self).zip(IntoIter::new(rhs)).zip(&mut dst) {
+            dst.write((lhs, rhs));
+        }
+        // FIXME: Convert to crate::mem::transmute once it works with generics.
+        // unsafe { crate::mem::transmute::<[MaybeUninit<U>; N], [U; N]>(dst) }
+        // SAFETY: At this point we've properly initialized the whole array
+        // and we just need to cast it to the correct type.
+        unsafe { crate::mem::transmute_copy::<_, [(T, U); N]>(&dst) }
```

Comment:
This implementation (x86-64 codegen with -O) uses more stack space and includes branches/a loop compared to this one (which is branch-free and uses utilizes less stack):

```rust
let mut dst = MaybeUninit::<[(T, U); N]>::uninit();
let ptr = dst.as_mut_ptr() as *mut (T, U);
for (idx, (lhs, rhs)) in IntoIter::new(lhs).zip(IntoIter::new(rhs)).enumerate() {
    unsafe { ptr.add(idx).write((lhs, rhs)) }
}
unsafe { dst.assume_init() }
```

https://godbolt.org/z/T4aT78

Reply:
However, that codegen is still not optimal.

Reply:
Also are `[MaybeUninit<(T, U)>; N]` and `[(T, U); N]` layout compatible/equivalent? Otherwise that transmute_copy would be UB.

Reply:
I am quite new to working with unsafe but the doc states
> [...] MaybeUninit<T> will always guarantee that it has the same size, alignment, and ABI as T [...]

and there are examples with [`[Vec<u32>; N]`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#initializing-an-array-element-by-element)) and `[u8; N]`. However I am not sure if there is anything special with `[(T, U); N]` that would make it not work.

Reply:
Looking at `map()` from #75212, which I based my function on, uses `unsafe { crate::mem::transmute_copy::<_, [U; N]>(&dst) }` for any `U`. That would make the same for `[(T, U); N]` (not the same `U`) fine, right?

Reply:
Should I change to your suggestion in https://github.com/rust-lang/rust/pull/79451#discussion_r531288075 or stay with my current code to stay more consistent with `map()`? (I am fine with either)

Reply:
The docs further says:

> However remember that a type containing a MaybeUninit<T> is not necessarily the same layout;

But in this case, I think it's safe because the concrete type here is an array.

> Should I change to your suggestion in #79451 (comment) or stay with my current code to stay more consistent with map()? (I am fine with either)

Personally at this point, I'd go with the current version, simply because it uses a single unsafe block compared to the suggested one above and we haven't done any benchmarks to prove that the suggestion above is actually significantly faster than the proposed one.

Reply:
We should benchmark it locally just to get an idea of the benefits and maybe apply them to other functions if possible and worth the code-complexity cost

Reply:
I have about zero experience bencmarking rust, however here is an attempt to benchmark the difference [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a7e556752ddd619b301b61365be10c28) with the following [result](https://gist.github.com/usbalbin/0ab4431f812db1824df90e480cafc129)

To me, on my 7 seven year old laptop, it seems like the speed is about the same. The original version even seems slightly faster on small inputs. However as I said, maybe I am bench marking wrong or the compiler is optimizing away something it would not be able to in a realistic example.

Reply:
Here is the test run with with lowest variance:
```
test tests::bench_zip         ... bench:          27 ns/iter (+/- 1)
test tests::bench_zip2        ... bench:          32 ns/iter (+/- 0)
test tests::bench_zip2_bigger ... bench:       5,613 ns/iter (+/- 142)
test tests::bench_zip_bigger  ... bench:       5,549 ns/iter (+/- 144)
```

where `zip` is my original implementation and `zip2` is @cynecx version

Reply:
No! Sorry I swapped them, `zip2` is mine

Reply:
@cynecx and @poliorcetics Sorry about the mixup. What do you think about the results? Is the speedup worth the extra `unsafe` block?

Reply:
If my math is right, the speed increase of @cynecx code is about 10-19% for small inputs

Reply:
There is also a stack size increase in the faster version, which I'm more concerned about because iterators are used even in embedded

Reply:
Ok, I'll stay with the current implementation for now

Reply:
> There is also a stack size increase in the faster version, which I'm more concerned about because iterators are used even in embedded

I think you may have swapped them. `bench_zip` and `bench_zip_bigger` (the one I suggested) are "faster" and use less stack space (~15%) according to the generated x86-64 code.

Reply:
> I think you may have swapped them. bench_zip and bench_zip_bigger (the one I suggested) are "faster" and use less stack space (~5-15%) according to the generated x86-64 code.

Ah ! My bad then, I'm certainly not against it in this case, especially if the unsafe blocks are well documented.

The standard library is by necessity full of unsafe, might as well put that to good use and optimize both speed and size when possible.

Reply:
I took a look at the llvm-ir again which basically boils down to (pseudocode):

```
zip(rhs: u8*, lhs: u8*, ret: u8*):
    iter = stackalloc(128)
    dst = stackalloc(128)

    memcpy(iter, lhs, 64)
    memcpy(iter + 128, rhs, 64)

    fill `iter` with `dst` in "zipped form"

    memcpy(ret, iter)
```

Theoretically, all three memcpys could be eliminated by copying the items directly from `rhs` and `lhs` into `ret`.

Reply:
@poliorcetics As I said before, I don't think the performance benefits are significant, so keeping the current version with one less unsafe block seems good enough for me.

Reply:
I'm personally more interested in the stack size reduction

Reply:
@poliorcetics If you look at the llvm-ir and the pseudocode above, llvm should've been able to eliminate the first two memcpys which would have reduce the stack usage by half. So it's more or less an llvm codegen issue.

Reply:
https://github.com/rust-lang/rust/issues/79754

Reply:
It would be useful to post links to implementations in godbolt for both to compare, as I see cynecx posted one but none for the original one. I think being able to quickly look at LLVM & asm of two implementations will help decide which one is more desired.

scottmcm also had some good suggestions for [optimizing map further](https://github.com/rust-lang/rust/pull/75571) (altho have been a bit busy lately so haven't had a chance to update that PR).

Reply:
Sorry for the delay. I added a `zippy2` to @cynecx's godbolt link [here](https://godbolt.org/z/Mdh7Mf). However I am not too familiar with judging asm quality :)

Reply:
ah ty for the update! The asm just helps me see what differences there are between the two versions, mainly looking for bounds checks which shouldn't be necessary here. The version you have has a loop in the LBB1 block, whereas cynecx's has no loop and is unrolled. If you increase the size of the arrays, you'll start to see more differences between the two.

The one thing is, if you change
```
for ((lhs, rhs), dst) in IntoIter::new(lhs).zip(IntoIter::new(rhs)).zip(&mut dst) {
  dst.write((lhs, rhs));
}
```
into
```
for (i, (lhs, rhs)) in IntoIter::new(lhs).zip(IntoIter::new(rhs)).enumerate() {
  dst[i].write((lhs, rhs));
}
```
both versions output the same asm. So the main difference between the two is here if you choose to zip or enumerate, rather than whether you use a pointer or a `MaybeUninit`

Reply:
Then would that be the preferred implementation (zip3 on [godbolt](https://godbolt.org/z/jnxMfh))? Seems to me like best of both worlds, no extra `unsafe` yet the same asm?

Reply:
To me that seems fine, but I think that's ultimately up to you & the reviewer! I was just curious how each impl compared to the others.

Reply:
Do you all think `zip3` is the way to go? Or do you have any other suggestions? :)